### PR TITLE
perf: Remove string allocations from method Colors.FromArbg

### PR DESCRIPTION
**Pull request by spouliot**
GitHub Issue (If applicable): closes #

## PR Type

What kind of change does this PR introduce?

- Refactoring (no functional changes, no api changes)

## What is the current behavior?

`Colors.FromArgb` is allocating several strings while parsing.

```
BenchmarkDotNet=v0.13.4, OS=macOS 13.0.1 (22A400) [Darwin 22.1.0]
Apple M1, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 7.0.0 (7.0.22.51805), Arm64 RyuJIT AdvSIMD

|   Method |    Mean |    Error |   StdDev |
|--------- |--------:|---------:|---------:|
| FromARGB | 5.724 s | 0.0118 s | 0.0110 s |
```

## What is the new behavior?

`Colors.FromArgb` does not allocate any temporary string while parsing.

```
BenchmarkDotNet=v0.13.4, OS=macOS 13.0.1 (22A400) [Darwin 22.1.0]
Apple M1, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 7.0.0 (7.0.22.51805), Arm64 RyuJIT AdvSIMD

|   Method |     Mean |   Error |  StdDev |
|--------- |---------:|--------:|--------:|
| FromARGB | 506.2 ms | 0.15 ms | 0.13 ms |
```

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Benchmark code:
https://github.com/spouliot/benchmarks/tree/main/Colors_FromARGB

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UWP/UI/Colors.cs [hunk 1]
```diff
@@ -71,45 +71,62 @@ public static Color Parse(string colorCode)
...
 
-			if (colorCode.Length == 3)
+			// deal with optional alpha value
+			if (len == 4)
 			{
-				a = 0xFF;
-				r = Convert.ToByte(new String(colorCode[0], 2), 16);
-				g = Convert.ToByte(new String(colorCode[1], 2), 16);
-				b = Convert.ToByte(new String(colorCode[2], 2), 16);
+				a = Convert.ToByte(colorCode[offset++]);
```
Comment (Youssef1313) [inline src/Uno.UWP/UI/Colors.cs:93]: This will produce the ASCII equivalent of the character. I don't think this is the intended behavior.

Comment (spouliot) [reply src/Uno.UWP/UI/Colors.cs]: nope, it was not :)

File: src/Uno.UWP/UI/Colors.cs [hunk 2]
```diff
@@ -71,45 +71,82 @@ public static Color Parse(string colorCode)
...
-				a = 0xFF;
 				r = 0xFF;
 				g = 0x0;
 				b = 0x0;
 			}
-			return Color.FromArgb(a, r, g, b);
+			return new Color(a, r, g, b);
+		}
+
+		static private byte ToByte(char c)
```
Comment (Youssef1313) [inline src/Uno.UWP/UI/Colors.cs:132]: Usually we put accessibility first.

```suggestion
		private static byte ToByte(char c)
```

Comment (Youssef1313) [inline src/Uno.UWP/UI/Colors.cs:132]: Usually we put accessibility first.

```suggestion
		private static byte ToByte(char c)
```

File: src/Uno.UWP/UI/Colors.cs [hunk 3]
```diff
@@ -71,45 +71,82 @@ public static Color Parse(string colorCode)
...
+			else
+			{
+				len = colorCode.Length;
+				// skip a starting `#` if present 
+				offset = (len > 0 && colorCode[0] == '#' ? 1 : 0);
+				len -= offset;
+			}
 
-			if (colorCode.Length == 3)
+			// deal with optional alpha value
```
Comment (agneszitte-nventive) [inline src/Uno.UWP/UI/Colors.cs:90]:
```suggestion
			// deal with an optional alpha value
```

File: src/Uno.UWP/UI/Colors.cs [hunk 4]
```diff
@@ -71,45 +71,82 @@ public static Color Parse(string colorCode)
...
+			else
 			{
 				a = 0xFF;
-				r = Convert.ToByte(colorCode.Substring(0, 2), 16);
-				g = Convert.ToByte(colorCode.Substring(2, 2), 16);
-				b = Convert.ToByte(colorCode.Substring(4, 2), 16);
 			}
-			else if (colorCode.Length == 8)
+
+			// the process the required R G and B values
```
Comment (jeromelaban) [inline src/Uno.UWP/UI/Colors.cs:107]:
```suggestion
			// then process the required R G and B values
```

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UWP/UI/Colors.cs [hunk 5]
```diff
@@ -71,45 +71,82 @@ public static Color Parse(string colorCode)
...
+				r = (byte)((ToByte(colorCode[offset++]) << 4) + ToByte(colorCode[offset++]));
+				g = (byte)((ToByte(colorCode[offset++]) << 4) + ToByte(colorCode[offset++]));
+				b = (byte)((ToByte(colorCode[offset++]) << 4) + ToByte(colorCode[offset++]));
 			}
 			else
 			{
-				a = 0xFF;
 				r = 0xFF;
 				g = 0x0;
 				b = 0x0;
```
Comment (Youssef1313) [inline src/Uno.UWP/UI/Colors.cs:127]: This looks like an error code path? I guess we probably should be throwing here instead of returning "Red" randomly here. This can be in another PR since this PR is focused on refactoring to improve performance without behavioral changes.

Comment (spouliot) [reply src/Uno.UWP/UI/Colors.cs:127]: Yes, that's the original behaviour.

From previous experience (silverlight/moonlight) using a color was more useful than an exception to debug color issues, since an exception was very hard to map back to the code that generated the exception. Not sure if this applies here... but that can be discussed on an issue/PR :)
